contrib/node/src/python/pants/contrib/node/subsystems/resolvers/npm_resolver.py

```diff
@@ -98,15 +100,60 @@ def resolve_target(self, node_task, target, results_dir, node_paths, resolve_loc
-                        .format(target.address.reference(), command, result))
+      # Install all dependencies except for `file:` dependencies
+      # `file:` dependencies are special dependencies that point to a local path to a pants node target
+      # `file:` dependencies are already in the build graph and should be already be installed by this point
+
+      # Copy the package.json and then remove the file: dependencies from package.json
+      # Run the install and symlink the file: dependencies using their node_paths
+      # Afterwards, restore the original package.json to not cause diff changes when resolve_locally=True
+      with safe_temp_edit('package.json') as package_json:
+        with open(package_json, 'r') as package_json_file:
+          json_data = json.load(package_json_file)
+          source_deps = { k : v for k,v in json_data.get('dependencies', {}).items() if v.startswith('file:')}
+          third_party_deps = { k : v for k,v in json_data.get('dependencies', {}).items() if not v.startswith('file:')}
+          json_data['dependencies'] = third_party_deps
+
+        # TODO(6489): Currently the file: dependencies need to be duplicated in BUILD.
+        # After this issue is closed, only dependencies need to be specified in package.json
+        for package_name, file_path in source_deps.items():
+          if self._get_target_from_package_name(target, package_name, file_path) is None:
+            raise TaskError('Local dependency in package.json not found in the build graph. '
+                            'Check your BUILD file for missing dependencies. ["{}": {}]'.format(package_name, file_path))
+        mode = 'w' if PY3 else 'wb'
+        with open(package_json, mode) as package_json_file:
+          json.dump(json_data, package_json_file, indent=2, separators=(',', ': '))
+        result, command = node_task.install_module(
+          target=target, install_optional=install_optional,
+          production_only=production_only, force=force, frozen_lockfile=frozen_lockfile,
+          workunit_name=target.address.reference(),
+          workunit_labels=[WorkUnitLabel.COMPILER])
+        if result != 0:
+          raise TaskError('Failed to resolve dependencies for {}:\n\t{} failed with exit code {}'
+                          .format(target.address.reference(), command, result))
+        if source_deps:
+          self._link_source_dependencies(target, results_dir, node_paths, source_deps)
+
+  def _link_source_dependencies(self, target, results_dir, node_paths, source_deps):
+    for package_name, file_path in source_deps.items():
+      dep = self._get_target_from_package_name(target, package_name, file_path)
+      # Symlink each target
+      dep_path = node_paths.node_path(dep)
+      node_module_dir = os.path.join(results_dir, 'node_modules')
+      absolute_symlink(dep_path, os.path.join(node_module_dir, dep.package_name))
```

Comment:
I think you want to use relative symlinks rather than absolute symlinks for cases like this... an absolute symlink will break if a git checkout is moved.